### PR TITLE
Implement IndicesRequest.Replaceable in RestoreSnapshotRequest

### LINKED ISSUE #21491: Implement IndicesRequest in RestoreSnapshotRequest
**Describe the feature**:

The restore snapshot API does allow to select indices that should be restored the same way the create snapshot does. Unfortunately the `org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotRequest` does not implement the `org.elasticsearch.action.IndicesRequest.Replaceable` interface like `org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest` does, although all needed methods are already present.

This prevents me currently from handling `RestoreSnapshotRequests` as `IndicesRequest`.

### PR DESCRIPTION
**Pull request by breml**
The restore snapshot API does allow to select indices that should be restored the same way the create snapshot does. Unfortunately the `org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotRequest` does not implement the `org.elasticsearch.action.IndicesRequest.Replaceable` interface like `org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest` does, although all needed methods are already present.

Closes #21491

### CODE REVIEW

**Comment by clintongormley** [comment]
@abeyad could you review please?

**Comment by breml** [comment]
@abeyad do you have any feedback for me on this PR?

**Review by abeyad — CHANGES_REQUESTED** [review]
I apologize for the delay, not sure how I missed the ping to review this.  It looks good, but the `indices(...)` method in `RestoreSnapshotRequest` should have the `@Override` annotation to indicate it overrides the interface method in `IndicesRequest.Replacable`.

Once that is done, I'll approve the PR and it can be merged.

Thanks for doing this!

**Comment by javanna** [comment]
sorry @abeyad @breml I had this on my radar too. This change doesn’t make much sense to me.  Why do you need to handle `RestoreSnapshotRequest` as a replaceable `IndicesRequest`?

**Comment by abeyad** [comment]
On second thought, I don't think this request object should actually implement IndicesRequest.  Its not of the same class of requests.

**Comment by breml** [comment]
@javanna @abeyad I'm aware of the fact that in case of a `RestoreSnapshotRequest` the indices are not indices within the Elasticsearch cluster but indices present in a snapshots. And therefore I understand your reaction, that this is not the same class of request. In principle I agree.

Nevertheless everything needed to implement the `IndicesRequests` interface is already present and it would simplify the implementation of my use case, outlined on the following lines.

Actual use case:

In the privileges filter of searchguard currently all operations of type `cluster:admin/snapshot/restore` are only allowed for users with administrator privileges. The reason for this limitation is, the breach of the security, if a non-administrator user is able to restore the index, in which the security privileges are stored (normally this would be the index called `searchguard`), because this could easily lead to privileges escalation. In my case I have the need, to be able to let non-administrator users to restore their indices (the indices, which they are granted access), so they could perform backups and restores of their indices on their own.

To make this happen, it would require to check two conditions:

1. Ensure, that the target index (defined by `rename_pattern` and `rename_replacement`) is allowed for the user.
2. Ensure, that the source index (in this case the index name in the snapshot on disk) is allowed for the user.

For the second part, searchguard already has the 	necessary logic implemented, which currently accepts arguments of type `IndicesRequest`.

Bottom line: if `RestoreSnapshotRequest` would implement the `IndicesRequest` interface, this would simplify the needed implementation for the above outlined use case. I hope this helps to clarify my motivation behind this PR.

The relevant code, which does the resolving of the index names and which uses IndicesRequests is here: https://github.com/floragunncom/search-guard/blob/master/src/main/java/com/floragunn/searchguard/configuration/PrivilegesEvaluator.java#L686

**Comment by javanna** [comment]
Hi @breml thanks for the thorough explanation.

`IndicesRequest` was designed to generically handle requests that relate to multiple indices, subject to name resolution based on cluster state. Like I said before, `RestoreSnapshotRequest` is a very different beast: its source indices get resolved against the indices present in the snapshot, and a `rename_pattern` is applied for the target indices, as you mentioned.

Using this interface would imply the promise that snapshot/restore indices can be dealt with in the same way as indices in the cluster, which is blatantly wrong. In other words, we will definitely not do this because it makes a promise about the way the code should work, and it doesn’t work in that way.

You already have specific code to deal with the target index and resolve the `rename_pattern`, I suggest to add the resolution of source indices as well to that without the need for the generic interface, rather casting the specific request to its own type that has its own special properties.
